include/boost/fusion/container/vector/detail/as_vector.hpp

```diff
@@ -8,12 +8,64 @@
 #define FUSION_AS_VECTOR_11052014_1801
 
 #include <boost/fusion/support/config.hpp>
+#include <boost/fusion/container/vector/detail/config.hpp>
 
 ///////////////////////////////////////////////////////////////////////////////
 // Without variadics, we will use the PP version
 ///////////////////////////////////////////////////////////////////////////////
+#if !defined(BOOST_FUSION_HAS_VARIADIC_VECTOR)
 # include <boost/fusion/container/vector/detail/cpp03/as_vector.hpp>
+#else
 
+///////////////////////////////////////////////////////////////////////////////
+// C++11 interface
+///////////////////////////////////////////////////////////////////////////////
+#include <boost/fusion/support/detail/index_sequence.hpp>
+#include <boost/fusion/container/vector/vector.hpp>
+#include <boost/fusion/iterator/value_of.hpp>
+#include <boost/fusion/iterator/deref.hpp>
+#include <boost/fusion/iterator/advance.hpp>
+#include <cstddef>
+
+namespace boost { namespace fusion { namespace detail
+{
+BOOST_FUSION_BARRIER_BEGIN
+
+    template <typename Indices>
+    struct as_vector_impl;
+
+    template <std::size_t ...Indices>
+    struct as_vector_impl<index_sequence<Indices...> >
+    {
+        template <typename Iterator>
+        struct apply
+        {
+            typedef vector<
+                numbered_vector_tag<sizeof...(Indices)>
+              , typename result_of::value_of<
+                    typename result_of::advance_c<Iterator, Indices>::type
+                >::type...
+            > type;
+        };
+
+        template <typename Iterator>
+        BOOST_CONSTEXPR BOOST_FUSION_GPU_ENABLED
+        static typename apply<Iterator>::type
+        call(Iterator i)
```

Comment:
I'd suggest against it. The result type has to be computed anyways, so this would create an unnecessary dependency on C++14, as well as creating a compilation time worry as "incomplete" functions need to be instantiated as soon as anyone touches them.